cpp/src/arrow/util/future.h

```diff
@@ -0,0 +1,318 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <atomic>
+#include <cmath>
+#include <memory>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "arrow/result.h"
+#include "arrow/status.h"
+#include "arrow/util/macros.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+enum class FutureState : int8_t { PENDING, SUCCESS, FAILURE };
+
+inline bool IsFutureFinished(FutureState state) { return state != FutureState::PENDING; }
+
+// ---------------------------------------------------------------------
+// Type-erased helpers
+
+class FutureWaiter;
+template <typename T>
+class Future;
+
+class ARROW_EXPORT FutureImpl {
+ public:
+  static constexpr double kInfinity = HUGE_VAL;
+
+  virtual ~FutureImpl();
+
+  FutureState state() { return state_.load(); }
+
+  static std::unique_ptr<FutureImpl> Make();
+
+ protected:
+  FutureImpl();
+  ARROW_DISALLOW_COPY_AND_ASSIGN(FutureImpl);
+
+  // Future API
+  void MarkFinished();
+  void MarkFailed();
+  void Wait();
+  bool Wait(double seconds);
+
+  // Waiter API
+  inline FutureState AddWaiter(FutureWaiter* w, int future_num);
```

Comment:
Since each future may only have one waiter, maybe this should be named `SetWaiter`? Or `AddToWaiter`?

Reply:
Good point, will do.